File: Frontend/Android/app/src/main/java/wiki/qaq/cookey/crypto/RequestAuthenticator.kt
package wiki.qaq.cookey.crypto

import android.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object RequestAuthenticator {

    fun computeRequestProof(
        rid: String,
        serverURL: String,
        targetURL: String,
        recipientPublicKey: String,
        deviceID: String,
        requestType: String,
        expiresAt: String,
        requestSecret: ByteArray
    ): String {
        val message = listOf(
            "cookey-request-v1",
            rid,
            serverURL,
            targetURL,
            recipientPublicKey,
            deviceID,
            requestType,
            expiresAt
        ).joinToString("\n")
        return hmacSha256Base64Url(requestSecret, message.toByteArray(Charsets.UTF_8))
    }

    fun verifyRequestProof(
        rid: String,
        serverURL: String,
        targetURL: String,
        recipientPublicKey: String,
        deviceID: String,
        requestType: String,
        expiresAt: String,
        requestSecret: ByteArray,
        expectedProof: String
    ): Boolean {
        val computed = computeRequestProof(
            rid, serverURL, targetURL, recipientPublicKey,
            deviceID, requestType, expiresAt, requestSecret
        )
        return computed == expectedProof
    }

    fun computeEnvelopeProof(
        rid: String,
        algorithm: String,
        ephemeralPublicKey: String,
        nonce: String,
        ciphertext: String,
        capturedAt: String,
        version: Int,
        requestSecret: ByteArray
    ): String {
        val message = listOf(
            "cookey-session-v1",
            rid,
            algorithm,
            ephemeralPublicKey,
            nonce,
            ciphertext,
            capturedAt,
            version.toString()
        ).joinToString("\n")
        return hmacSha256Base64Url(requestSecret, message.toByteArray(Charsets.UTF_8))
    }

    private fun hmacSha256Base64Url(key: ByteArray, data: ByteArray): String {
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(key, "HmacSHA256"))
        val result = mac.doFinal(data)
        return Base64.encodeToString(result, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    }
}

File: Frontend/Apple/Cookey/Services/RequestAuthenticator.swift
import CryptoKit
import Foundation

enum RequestAuthenticator {
    enum Error: LocalizedError {
        case invalidRequestSecret
        case invalidRequestProof
        case invalidSessionProof
        case missingAuthenticatedFields

        var errorDescription: String? {
            switch self {
            case .invalidRequestSecret:
                "Invalid request secret."
            case .invalidRequestProof:
                "The request could not be authenticated."
            case .invalidSessionProof:
                "The uploaded session proof was invalid."
            case .missingAuthenticatedFields:
                "The request is missing authenticated fields."
            }
        }
    }

    static func verify(_ deepLink: DeepLink) throws {
        guard
            let requestSecret = deepLink.requestSecret,
            let providedRequestProof = deepLink.requestProof,
            let expiresAt = deepLink.expiresAt
        else {
            throw Error.missingAuthenticatedFields
        }

        let expected = try Self.requestProof(
            rid: deepLink.rid,
            serverURL: deepLink.serverURL,
            targetURL: deepLink.targetURL,
            recipientPublicKeyBase64: deepLink.recipientPublicKeyBase64,
            deviceID: deepLink.deviceID,
            requestType: deepLink.requestType,
            expiresAt: expiresAt,
            requestSecret: requestSecret,
        )

        guard expected == providedRequestProof else {
            throw Error.invalidRequestProof
        }
    }

    static func requestProof(
        rid: String,
        serverURL: URL,
        targetURL: URL,
        recipientPublicKeyBase64: String,
        deviceID: String,
        requestType: DeepLink.RequestType,
        expiresAt: Date,
        requestSecret: String,
    ) throws -> String {
        try authenticate(
            purpose: "cookey-request-v1",
            secret: requestSecret,
            fields: [
                rid,
                serverURL.absoluteString,
                targetURL.absoluteString,
                recipientPublicKeyBase64,
                deviceID,
                requestType.rawValue,
                timestamp(expiresAt),
            ],
        )
    }

    static func envelopeProof(
        rid: String,
        envelope: EncryptedSessionEnvelope,
        requestSecret: String,
    ) throws -> String {
        try authenticate(
            purpose: "cookey-session-v1",
            secret: requestSecret,
            fields: [
                rid,
                envelope.algorithm,
                envelope.ephemeralPublicKey,
                envelope.nonce,
                envelope.ciphertext,
                timestamp(envelope.capturedAt),
                String(envelope.version),
            ],
        )
    }

    private static func authenticate(
        purpose: String,
        secret: String,
        fields: [String],
    ) throws -> String {
        guard let secretData = Data(base64URLEncoded: secret), secretData.count >= 16 else {
            throw Error.invalidRequestSecret
        }

        let message = ([purpose] + fields).joined(separator: "\n")
        let mac = HMAC<SHA256>.authenticationCode(
            for: Data(message.utf8),
            using: SymmetricKey(data: secretData),
        )
        return Data(mac).base64URLEncodedString()
    }

    private static func timestamp(_ date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: date)
    }
}

private extension Data {
    init?(base64URLEncoded string: String) {
        let normalized = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let padding = (4 - normalized.count % 4) % 4
        self.init(base64Encoded: normalized + String(repeating: "=", count: padding))
    }

    func base64URLEncodedString() -> String {
        base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
